controllers/automation/policyautomation_controller.go

```diff
@@ -89,6 +92,150 @@ func (r *PolicyAutomationReconciler) setOwnerReferences(
+			hostAddr, err := net.LookupAddr(hostStr)
+			if err == nil {
+				hubCluster = hostAddr[0]
+			}
+		}
+	}
+
+	log.V(3).Info("Hub Cluster information: %s", hubCluster)
+
+	return hubCluster
+}
+
+// getViolationContext will put the root policy information into violationContext
+// It also put the status of the non-compliance replicated policies into violationContext
+func (r *PolicyAutomationReconciler) getViolationContext(
+	policy *policyv1.Policy,
+	targetList []string,
+	policyAutomation *policyv1beta1.PolicyAutomation,
+) (policyv1beta1.ViolationContext, error) {
+	log.V(3).Info(
+		"Get the violation context from the root policy %s/%s",
+		policy.GetNamespace(),
+		policy.GetName(),
+	)
+
+	violationContext := policyv1beta1.ViolationContext{}
+	// 1) get the target cluster list
+	violationContext.TargetClusters = targetList
+	// 2) get the root policy name
+	violationContext.PolicyName = policy.GetName()
+	// 3) get the root policy namespace
+	violationContext.PolicyNamespace = policy.GetNamespace()
+	// 4) get the root policy hub cluster name
+	violationContext.HubCluster = getClusterName()
+
+	// 5) get the policy sets of the root policy
+	plcPlacement := policy.Status.Placement
+	policySets := []string{}
+
+	for _, placement := range plcPlacement {
+		if placement.PolicySet != "" {
+			policySets = append(policySets, placement.PolicySet)
+		}
+	}
+
+	violationContext.PolicySets = policySets
+
+	replicatedPlcList := &policyv1.PolicyList{}
+
+	err := r.List(
+		context.TODO(),
+		replicatedPlcList,
+		client.MatchingLabels(common.LabelsForRootPolicy(policy)),
+	)
+	if err != nil {
+		log.Error(err, "Failed to list the replicated policies")
+
+		return violationContext, err
+	}
+
+	if len(replicatedPlcList.Items) == 0 {
+		log.V(2).Info("The replicated policies cannot be found.")
+
+		return violationContext, nil
+	}
+
+	policyViolationContextLimit := policyAutomation.Spec.Automation.PolicyViolationContextLimit
+	if policyViolationContextLimit == nil {
+		policyViolationContextLimit = new(uint)
+		*policyViolationContextLimit = policyv1beta1.DefaultPolicyViolationContextLimit
+	}
+
+	contextLimit := int(*policyViolationContextLimit)
+
+	targetListMap := getTargetListMap(targetList)
+	violationContext.PolicyViolationContext = make(
+		map[string]policyv1beta1.ReplicatedPolicyStatus,
+		len(replicatedPlcList.Items),
+	)
+
+	// 6) get the status of the non-compliance replicated policies
+	for _, rPlc := range replicatedPlcList.Items {
+		clusterName := rPlc.GetLabels()[common.ClusterNameLabel]
+		if !targetListMap[clusterName] {
+			continue // skip the compliance replicated policies
+		}
+
+		rPlcStatus := policyv1beta1.ReplicatedPolicyStatus{}
+		// Convert PolicyStatus to ReplicatedPolicyStatus and skip the unnecessary items
+		err := common.TypeConverter(rPlc.Status, &rPlcStatus)
+		if err != nil {
+			log.Error(err, "The PolicyStatus cannot be converted to the type ReplicatedPolicyStatus.")
+
+			continue
```

Comment:
I'm actually thinking an empty `rPlcStatus` would be better than skipping it entirely.